test/qa/test_meshnet.py

```diff
@@ -170,3 +196,120 @@ def test_set_meshnet_off_repeated(meshnet_allias):
             sh.nordvpn.set(meshnet_allias, "off")
 
     assert "Meshnet is already disabled." in str(ex.value)
+
+
+@pytest.mark.parametrize(("permission", "permission_state", "expected_message"), meshnet.PERMISSION_SUCCESS_MESSAGE_PARAMETER_SET, \
+                         ids=[f"{line[0]}-{line[1]}" for line in meshnet.PERMISSION_SUCCESS_MESSAGE_PARAMETER_SET])
+@timeout_decorator.timeout(25)
+def test_permission_messages_success(permission, permission_state, expected_message):
+    peer_hostname = meshnet.PeerList.from_str(sh.nordvpn.mesh.peer.list()).get_external_peer().hostname
+
+    reverse_permission_value = "allow" if permission_state == "deny" else "deny"
+    meshnet.set_permission(peer_hostname, permission, reverse_permission_value)
+
+    got_message = sh.nordvpn.mesh.peer(permission, permission_state, peer_hostname)
+
+    expected_message = expected_message % peer_hostname
+
+    assert expected_message in got_message
+
+
+@pytest.mark.parametrize(("permission", "permission_state", "expected_message"), meshnet.PERMISSION_ERROR_MESSAGE_PARAMETER_SET, \
+                         ids=[f"{line[0]}-{line[1]}" for line in meshnet.PERMISSION_ERROR_MESSAGE_PARAMETER_SET])
+@timeout_decorator.timeout(25)
+def test_permission_messages_error(permission, permission_state, expected_message):
+    peer_hostname = meshnet.PeerList.from_str(sh.nordvpn.mesh.peer.list()).get_external_peer().hostname
+
+    sh.nordvpn.mesh.peer(permission, permission_state, peer_hostname, _ok_code=(0, 1))
+
+    with pytest.raises(sh.ErrorReturnCode_1) as ex:
+        print(sh.nordvpn.mesh.peer(permission, permission_state, peer_hostname))
+
+    expected_message = expected_message % peer_hostname
+
+    assert expected_message in str(ex)
+
+
+def test_derp_server_selection_logic():
+    def has_duplicates(list):
+        return len(list) != len(set(list))
+
+    ssh_client.exec_command("sudo iptables -I OUTPUT 1 -p tcp -m tcp --sport 8765 -j DROP")
+    ssh_client.exec_command("sudo iptables -I OUTPUT 1 -p tcp -m tcp --dport 8765 -j DROP")
+
+    daemon.stop_peer(ssh_client)
+    ssh_client.exec_command("echo '' > /var/log/nordvpn/daemon.log")
+    daemon.start_peer(ssh_client)
+
+    time.sleep(30)
```

Comment:
why 30 seconds? :)

Reply:
Pretty much same test just for VPN connection is done @ test_connect.py::test_status_connected, so I thought we could keep the same logic :smile: